java-checks/src/main/java/org/sonar/java/checks/spring/SpringBeansShouldBeAccessibleCheck.java

```diff
@@ -0,0 +1,157 @@
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import org.sonar.check.Rule;
+import org.sonar.java.AnalyzerMessage;
+import org.sonar.java.AnalyzerMessageReporter;
+import org.sonar.java.CrossFileScanner;
+import org.sonar.java.checks.helpers.ConstantUtils;
+import org.sonar.java.model.PackageUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.tree.AnnotationTree;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.CompilationUnitTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.NewArrayTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S4605")
+public class SpringBeansShouldBeAccessibleCheck extends IssuableSubscriptionVisitor implements CrossFileScanner {
+
+  private static final String MESSAGE_FORMAT = "'%s' is not reachable by @ComponentsScan or @SpringBootApplication. "
+    + "Either move it to a package configured in @ComponentsScan or update your @ComponentsScan configuration.";
+
+  private static final String[] SPRING_BEAN_ANNOTATIONS = {
+    "org.springframework.stereotype.Component",
+    "org.springframework.stereotype.Service",
+    "org.springframework.stereotype.Repository",
+    "org.springframework.stereotype.Controller",
+    "org.springframework.web.bind.annotation.RestController"
+  };
+
+  private static final String COMPONENT_SCAN_ANNOTATION = "org.springframework.context.annotation.ComponentScan";
+  private static final Set<String> COMPONENT_SCAN_ARGUMENTS = new HashSet<>(Arrays.asList("basePackages", "value"));
+
+  private static final String SPRING_BOOT_APP_ANNOTATION = "org.springframework.boot.autoconfigure.SpringBootApplication";
+
+  private final Map<String, List<AnalyzerMessage>> messagesPerPackage = new HashMap<>();
+  private final Set<String> scannedPackages = new HashSet<>();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.CLASS);
+  }
+
+  @Override
+  public void endOfAnalysis() {
+    if (!(context instanceof AnalyzerMessageReporter)) {
+      return;
+    }
+
+    AnalyzerMessageReporter reporter = (AnalyzerMessageReporter) context;
+    messagesPerPackage.entrySet().stream()
+      // also consider sub-packages
+      .filter(entry -> scannedPackages.stream().noneMatch(entry.getKey()::contains))
+      .forEach(entry -> entry.getValue().forEach(reporter::reportIssue));
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic() || !(context instanceof AnalyzerMessageReporter)) {
+      return;
+    }
+
+    ClassTree classTree = (ClassTree) tree;
+    String classPackageName = packageNameOf(classTree);
+
+    Optional<AnnotationTree> componentScanAnnotation = classTree.modifiers().annotations().stream()
```

Comment:
I think it would be simpler to rewrite all this part relying on `valuesForAnnotation(String)` from the symbolMetdata, and then check for values.

If the method returns null, it is not annotated with the annotation, if it return a list, then you have directly the values (which could be both symbols or trees, but we probably only needs trees)